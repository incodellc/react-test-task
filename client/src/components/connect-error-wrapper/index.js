import ConnectErrorWrapper from './ConnectErrorWrapper';
export default ConnectErrorWrapper;