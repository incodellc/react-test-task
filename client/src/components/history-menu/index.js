import HistoryMenu from './HistoryMenu';
export default HistoryMenu;