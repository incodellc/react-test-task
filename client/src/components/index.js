import InfoBar from './info-bar';
import Layout from './layout';
import ConnectErrorWrapper from './connect-error-wrapper';
import TickerList from './ticker-list';
import TickerListItem from './ticker-list-item';

export { InfoBar, Layout, ConnectErrorWrapper, TickerList, TickerListItem };