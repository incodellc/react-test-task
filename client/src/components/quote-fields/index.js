import QuoteFields from './QuoteFields';
export default QuoteFields;