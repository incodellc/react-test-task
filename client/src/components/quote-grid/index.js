import QuotesGrid from './QuotesGrid';
export default QuotesGrid;