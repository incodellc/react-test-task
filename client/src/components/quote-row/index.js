import QuoteRow from './QuoteRow';
export default QuoteRow;