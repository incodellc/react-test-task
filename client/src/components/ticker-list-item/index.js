import TickerListItem from './TickerListItem';
export default TickerListItem;