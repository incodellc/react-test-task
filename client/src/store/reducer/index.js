import actionTypes from '../actions/action-types';
import { fields, tickers, historyLengthOptions, rewriteObjectProps } from '../../utils';

export const initState = {
    connecting: true,
    connected: false,
    fields: [...fields],
    tickers: [...tickers],
    bannedTickerIds: [],
    quotes: [],
    sortParams: {
        target: null,
        isDate: false,
        order: null,
        reverse: false
    },
    historyLength: 100,
    historyLengthOptions: [...historyLengthOptions]
};

export const reducer = (state = initState, action) => {
    switch (action.type) {

        case actionTypes.CONNECTION_INIT:
            return rewriteObjectProps(state, {
                connecting: true,
                bannedTickerIds: [],
                quotes: [],
                sort: rewriteObjectProps(state.sort, {
                    target: null,
                    isDate: false,
                    order: null,
                    reverse: false
                })
            });

        case actionTypes.STREAM_SERVER_CONNECT:
            if (!state.connected) {
                return rewriteObjectProps(state, {
                    connected: true,
                    connecting: false
                });
            }
            return state;

        case actionTypes.STREAM_SERVER_DISCONNECT:
            return rewriteObjectProps(state, {
                connected: false,
                connecting: false
            });
        
        case actionTypes.UPDATE_QUOTES:
            return rewriteObjectProps(state, {
                quotes: [action.data].concat(state.quotes.slice(0, state.historyLength - 1))
            });

        case actionTypes.TOGGLE_TICKER:
            {
                let index = state.bannedTickerIds.findIndex(item => item === action.tickerId);
                let bannedTickerIds = state.bannedTickerIds;
                let newBannedTickerIds = index > -1 
                    ? bannedTickerIds.slice(0, index).concat(bannedTickerIds.slice(index + 1, state.bannedTickerIds.length))
                    : bannedTickerIds.concat(action.tickerId);
                return rewriteObjectProps(state, {
                    bannedTickerIds: newBannedTickerIds
                });
            }

        case actionTypes.SET_HISTORY_LENGTH:
            return rewriteObjectProps(state, {
                historyLength: action.lengthValue
            });

        case actionTypes.APPLY_SORT_PARAMS:
            
        default:
            return state;

    }
};