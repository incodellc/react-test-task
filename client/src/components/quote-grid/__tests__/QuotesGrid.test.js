import React from 'react';
import { initState } from '../../../store/reducer';
import { quotesExample } from '../../../utils'
import { renderWithRedux } from '../../../utils/test-utils';
import { QuotesGrid } from '../../../components';
import { cleanup } from '@testing-library/react';

describe('Check QuotesGrid component', () => {
    afterEach(cleanup);

    it('component renders correctly', () => {
        const { asFragment } = renderWithRedux(<QuotesGrid />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('component should contain number of QuoteRow comp-s equals to all quotes', () => {
        const { getByTestId, getAllByTestId, store } = renderWithRedux(<QuotesGrid />);
        expect(getByTestId('quotes-grid')).toContainElement(getAllByTestId('quote-row')[0]);
        expect(getAllByTestId('quote-row').length).toEqual(store.getState().quotes.length);
    });

    it('component should contain number of QuoteRow comp-s equals to all quotes except two', () => {
        const state = { ...initState, quotes: quotesExample.slice(), bannedTickerIds: ['AAPL', 'EBAY'] };
        const { getAllByTestId, store } = renderWithRedux(<QuotesGrid />, state);
        expect(getAllByTestId('quote-row').length).toEqual(store.getState().quotes.length - 2);
    });

    it('component should be empty', () => {
        const state = { ...initState, quotes: [] };
        const { queryAllByTestId } = renderWithRedux(<QuotesGrid />, state);
        expect(queryAllByTestId('quote-row').length).toEqual(0);
    });
});