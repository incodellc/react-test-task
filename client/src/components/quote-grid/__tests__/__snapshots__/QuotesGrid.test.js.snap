// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check QuotesGrid component component renders correctly 1`] = `
<DocumentFragment>
  <table
    class="ui celled fixed selectable table quotesGrid"
    data-testid="quotes-grid"
  >
    <tbody
      class=""
    >
      <tr
        class="negative"
        data-testid="quote-row"
      >
        <td
          class=""
          data-testid="quote-cell"
        >
          NASDAQ
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          AAPL
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          <i
            aria-hidden="true"
            class="arrow alternate circle down icon"
            data-testid="quote-cell-icon-down"
          />
          213.41
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          -38.72
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          -0.18
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.56
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.55
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          15-02-20 23:40:17
        </td>
      </tr>
      <tr
        class="positive"
        data-testid="quote-row"
      >
        <td
          class=""
          data-testid="quote-cell"
        >
          NASDAQ
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          GOOGL
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          <i
            aria-hidden="true"
            class="arrow alternate circle up icon"
            data-testid="quote-cell-icon-up"
          />
          253.86
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          5.59
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.02
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.82
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          1.08
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          15-02-20 23:48:18
        </td>
      </tr>
      <tr
        class="positive"
        data-testid="quote-row"
      >
        <td
          class=""
          data-testid="quote-cell"
        >
          NASDAQ
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          AMZN
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          <i
            aria-hidden="true"
            class="arrow alternate circle up icon"
            data-testid="quote-cell-icon-up"
          />
          244.82
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          21.81
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.09
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.82
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          1.30
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          15-02-20 23:42:18
        </td>
      </tr>
      <tr
        class="positive"
        data-testid="quote-row"
      >
        <td
          class=""
          data-testid="quote-cell"
        >
          NASDAQ
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          ADBE
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          <i
            aria-hidden="true"
            class="arrow alternate circle up icon"
            data-testid="quote-cell-icon-up"
          />
          275.75
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          11.16
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.04
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.58
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          1.02
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          15-02-20 23:45:18
        </td>
      </tr>
      <tr
        class="negative"
        data-testid="quote-row"
      >
        <td
          class=""
          data-testid="quote-cell"
        >
          NASDAQ
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          CSCO
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          <i
            aria-hidden="true"
            class="arrow alternate circle down icon"
            data-testid="quote-cell-icon-down"
          />
          225.81
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          -12.73
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          -0.06
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.46
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          1.12
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          15-02-20 23:44:18
        </td>
      </tr>
      <tr
        class="positive"
        data-testid="quote-row"
      >
        <td
          class=""
          data-testid="quote-cell"
        >
          NASDAQ
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          DISCK
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          <i
            aria-hidden="true"
            class="arrow alternate circle up icon"
            data-testid="quote-cell-icon-up"
          />
          257.54
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          4.77
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.02
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.82
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          1.31
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          15-02-20 23:50:18
        </td>
      </tr>
      <tr
        class="negative"
        data-testid="quote-row"
      >
        <td
          class=""
          data-testid="quote-cell"
        >
          NASDAQ
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          EBAY
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          <i
            aria-hidden="true"
            class="arrow alternate circle down icon"
            data-testid="quote-cell-icon-down"
          />
          210.04
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          -29.38
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          -0.14
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.24
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.00
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          15-02-20 23:46:18
        </td>
      </tr>
      <tr
        class="negative"
        data-testid="quote-row"
      >
        <td
          class=""
          data-testid="quote-cell"
        >
          NASDAQ
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          NFLX
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          <i
            aria-hidden="true"
            class="arrow alternate circle down icon"
            data-testid="quote-cell-icon-down"
          />
          230.34
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          -13.58
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          -0.06
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.21
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.54
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          15-02-20 23:47:18
        </td>
      </tr>
      <tr
        class="positive"
        data-testid="quote-row"
      >
        <td
          class=""
          data-testid="quote-cell"
        >
          NASDAQ
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          MSFT
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          <i
            aria-hidden="true"
            class="arrow alternate circle up icon"
            data-testid="quote-cell-icon-up"
          />
          246.52
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          32.02
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          -0.13
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.75
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          1.05
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          15-02-20 23:49:18
        </td>
      </tr>
      <tr
        class="negative"
        data-testid="quote-row"
      >
        <td
          class=""
          data-testid="quote-cell"
        >
          NASDAQ
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          TSLA
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          <i
            aria-hidden="true"
            class="arrow alternate circle down icon"
            data-testid="quote-cell-icon-down"
          />
          204.50
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          -45.42
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          -0.22
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          0.63
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          1.11
        </td>
        <td
          class=""
          data-testid="quote-cell"
        >
          15-02-20 23:41:18
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
